### PR TITLE
WT-7102 Migrate full and incremental backup functionalities into wtbackup class

### PR DESCRIPTION
**Pull request by boboeee**
This ticket aims to improve on the refactoring of the backup classes, being a follow-up ticket from WT-6866. The functionality of **take_full_backup**, **take_incr_backup** and a new functionality called **take_log_backup** is now used as common code in the wtbackup class. The functionalities of what to return, and what to pass into the backup functions varied across different test files, thus the functions don't look exactly the same, and there were modifications need to be done, to allow commonality between different tests.

### CODE REVIEW

**Comment by boboeee** [comment]
hey @sueloverso  and @ddanderson this is ready for review now :)

File: test/suite/test_backup04.py [hunk 1]
```diff
@@ -76,37 +76,13 @@ def update(self, uri, dsize, upd, rows):
...
-                sz = os.path.getsize(newfile)
-                self.pr('Copy from: ' + newfile + ' (' + str(sz) + ') to ' + dir)
-                shutil.copy(newfile, dir)
-            self.assertEqual(ret, wiredtiger.WT_NOTFOUND)
-            self.session.truncate('log:', cursor, None, None)
-            cursor.close()
+    def take_incr_backup_and_truncate(self, dir):
+        config = 'target=("log:")'
+        cursor = self.session.open_cursor('backup:', None, config)
+        self.take_full_backup(dir, cursor)
```
Comment (boboeee) [inline test/suite/test_backup04.py:83]: This function might be abit ambiguous, but not quite sure how to name this @sueloverso any thoughts?

File: test/suite/test_backup13.py [hunk 2]
```diff
@@ -177,8 +124,7 @@ def test_backup13(self):
         bkup_c.close()
 
         # Make sure after a force stop we cannot access old backup info.
-        config = 'incremental=(src_id="ID1",this_id="ID3")'
-
+        config = 'incremental=(src_id="ID1",this_id="ID2")'
```
Comment (boboeee) [inline test/suite/test_backup13.py:127]: I don't think ID3 exist, so i changed this, such that we can make sure that we are checking access rights for old backups let me know If I should change this back

File: test/suite/test_backup17.py [hunk 3]
```diff
@@ -52,60 +52,19 @@ class test_backup17(backup_base):
...
-        if consolidate:
-            buf += ',consolidate=true'
-        buf += ')'
-        bkup_c = self.session.open_cursor('backup:', None, buf)
-        lens = []
+    #
+    # With a file length list, and given consolidate option, check
+    # that we get the appropriate lengths from incremental backups
+    #
+    def check_consolidate_sizes(self, file_lens, consolidate):
```
Comment (boboeee) [inline test/suite/test_backup17.py:59]: Hey @sueloverso  for this function, could there be a better way to explain what the function is doing?

File: test/suite/wtbackup.py [hunk 4]
```diff
@@ -92,22 +97,21 @@ def populate(self, objs, do_checkpoint=False, skiplsm=False):
     # iteration and an incremental backup for each iteration. That way we can compare the full and
     # incremental each time through.
     #
-    def setup_directories(self, max_iteration, home_incr, home_full, logpath):
-        for i in range(0, max_iteration):
-            # The log directory is a subdirectory of the home directory,
-            # creating that will make the home directory also.
+    def setup_directories(self, home_incr, home_full):
```
Comment (boboeee) [inline test/suite/wtbackup.py:100]: I have repurposed setup_directories, to now also setup directories for any future incremental backup and full backup directories. Let me know if there is a cleaner way to do this!

File: test/suite/wtbackup.py [hunk 5]
```diff
@@ -144,6 +192,143 @@ def compare_backups(self, uri, base_dir_home, other_dir_home, suffix = None):
...
+        while True:
+            ret = bkup_c.next()
+            if ret != 0:
+                break
+            newfile = bkup_c.get_key()
+            file_sizes = self.take_incr_backup_block(bkup_c, newfile, backup_incr_dir)
+            if ret_sizes:
+                files_info += file_sizes
+            else:
+                files_info.append(newfile)
```
Comment (ddanderson) [inline test/suite/wtbackup.py:329]: Having this function return two things, of different types, depending on a parameter, is hard to follow.
I think it is better to not have the `ret_sizes` parameter and always return multiple values (`return (file_names, file_sizes)`.  Caller does `(file_names, file_sizes) = take_incr_backup(...)` and you can choose to ignore info you don't need.

**Review by ddanderson — COMMENTED** [review]
@boboeee, I've noted a couple places that I think could become clearer to read.  Otherwise, there's a lot here and it's hard to verify so many changes.  I think that one way to confirm that large refactoring hasn't removed some important functionality may be as follows:
  1) make sure old (prefactored) code has print statements or some sort of logging. (For these tests, I think we have that in `self.pr` statements, which should generate output into `WT_TEST/results.txt`)
  2) refactor, keeping print statements (it appears you've done that)
  3) run complete set of tests and for each, compare output from `self.pr` before and after, making sure they are identical.  If you needed to add/subtract self.pr statements, of course, you need to adjust the comparison.

Have you done something like that?

File: test/suite/test_backup04.py (hunk 1 above)
Comment (sueloverso) [reply test/suite/test_backup04.py:83]: The truncate is a specific feature for log-based incremental backup. I would just call it `take_log_incr_backup`. I think it is find to leave the block-based one as `take_incr_backup` because that is the predominant type of backup we test and MongoDB uses.

File: test/suite/test_backup04.py [hunk 6]
```diff
@@ -118,8 +94,9 @@ def test_incremental_backup(self):
         self.populate_with_string(self.uri, self.dsize, self.nops)
```
Comment (sueloverso) [inline test/suite/test_backup04.py:94]: This function would be better named `test_log_incremental_backup`.

This test was written before block-based incremental backup existed. The comments in this file (and any others using log-based incremental) should be updated to explicitly say log-based incremental backup.

File: test/suite/test_backup13.py (hunk 2 above)
Comment (sueloverso) [reply test/suite/test_backup13.py:127]: You're correct `ID3` does not exist. The `this_id` parameter is supposed to be new. Both `ID1` and `ID2` have been used. So while this config should return an error due and not find `ID1` if it fails or something is wrong, we want to use an unused ID.

File: test/suite/test_backup17.py (hunk 3 above)
Comment (sueloverso) [reply test/suite/test_backup17.py:59]: Well, what it says is accurate. Maybe more detail would be helpful. Here's detail:
When the consolidate option is used, we expect the incremental backup to collapse adjacent blocks and return us lengths that exceed the granularity setting. Verify we see multiple blocks. If consolidate is not used, no block lengths should ever be greater than the granularity setting.

File: test/suite/wtbackup.py [hunk 7]
```diff
@@ -25,33 +25,38 @@
...
     # To determine whether to increase/decrease counter, which determines
-    initial_backup = True
+    initial_backup = False
     # Used for populate function
     rows = 100
     populate_big = None
 
+    # Specify a logpath directory to be used to place wiredtiger log files
+    logpath=''
+    # Temporary directory used to consistent data between multiple incremental backups
```
Comment (sueloverso) [inline test/suite/wtbackup.py:50]: Missing a word: "... used to verify consistent data..."

File: test/suite/wtbackup.py [hunk 8]
```diff
@@ -25,33 +25,38 @@
...
+    # of incremental backup.
+    bkup_id = 0
     # To determine whether to increase/decrease counter, which determines
-    initial_backup = True
+    initial_backup = False
     # Used for populate function
     rows = 100
     populate_big = None
 
+    # Specify a logpath directory to be used to place wiredtiger log files
```
Comment (sueloverso) [inline test/suite/wtbackup.py:48]: Here and other places, all comments should be full sentences, ending in a period.

**Review by sueloverso — COMMENTED** [review]
I have several comments here.

**Comment by boboeee** [comment]
Hey @ddanderson , thats a great idea! I still haven't aligned prints between all of the tests

File: test/suite/wtbackup.py [hunk 9]
```diff
@@ -144,6 +192,143 @@ def compare_backups(self, uri, base_dir_home, other_dir_home, suffix = None):
...
+        incr_c = self.session.open_cursor(None, bkup_c, config)
+        # For consolidate
+        lens = []
+        # We cannot use 'for newfile in incr_c:' usage because backup cursors don't have
+        # values and adding in get_values returns ENOTSUP and causes the usage to fail.
+        # If that changes then this, and the use of the duplicate below can change.
+        while True:
+            ret = incr_c.next()
+            if ret != 0:
+                break
```
[the remark this answers is not in the document]
Comment (boboeee) [reply test/suite/wtbackup.py:249]: ooo i like this! makes the code look very clean

File: test/suite/wtbackup.py (hunk 5 above)
Comment (boboeee) [reply test/suite/wtbackup.py:329]: done!

File: test/suite/test_backup04.py (hunk 6 above)
Comment (boboeee) [reply test/suite/test_backup04.py:94]: done

File: test/suite/test_backup13.py (hunk 2 above)
Comment (boboeee) [reply test/suite/test_backup13.py:127]: oh i see, will revert back thanks!

File: test/suite/test_backup17.py (hunk 3 above)
Comment (boboeee) [reply test/suite/test_backup17.py:59]: thanks ill add in that comment

File: test/suite/wtbackup.py (hunk 7 above)
Comment (boboeee) [reply test/suite/wtbackup.py:50]: thanks!

File: test/suite/wtbackup.py (hunk 8 above)
Comment (boboeee) [reply test/suite/wtbackup.py:48]: done!

File: test/suite/test_backup04.py (hunk 1 above)
Comment (boboeee) [reply test/suite/test_backup04.py:83]: awesome ta!

**Comment by ddanderson** [comment]
@boboeee That's great, it doesn't have to be perfect as long as you understand the reasons for any variance in output.  Thanks for doing that.

File: test/suite/wtbackup.py [hunk 10]
```diff
@@ -92,22 +97,21 @@ def populate(self, objs, do_checkpoint=False, skiplsm=False):
     # iteration and an incremental backup for each iteration. That way we can compare the full and
     # incremental each time through.
     #
-    def setup_directories(self, max_iteration, home_incr, home_full, logpath):
-        for i in range(0, max_iteration):
-            # The log directory is a subdirectory of the home directory,
-            # creating that will make the home directory also.
+    def setup_directories(self, home_incr, home_full):
+        if not os.path.exists(self.home_tmp):
+            os.makedirs(self.home_tmp + '/' + self.logpath)
```
Comment (ddanderson) [inline test/suite/wtbackup.py:102]: I think I would leave out the `if` part and do the `os.makedirs` unconditionally (`os.makedirs` creates dirs as needed).  It doesn't make much sense to me to say if "some_path" doesn't exist then make "some_path/log".  If for some reason you _don't_ want to create the logpath if the home_tmp already exists (and therefore need the original code), please explain with a comment.

File: test/suite/wtbackup.py [hunk 11]
```diff
@@ -92,22 +97,21 @@ def populate(self, objs, do_checkpoint=False, skiplsm=False):
...
-                continue
-            home_full_dir = home_full + '.' + str(i)
-            if os.path.exists(home_full_dir):
-                os.remove(home_full_dir)
-            os.makedirs(home_full_dir + '/' + logpath)
+        if os.path.exists(home_incr):
+            shutil.rmtree(home_incr)
+            shutil.copytree(self.home_tmp, self.home_incr)
+        else:
+            os.makedirs(home_incr + '/' + self.logpath)
```
Comment (ddanderson) [inline test/suite/wtbackup.py:114]: This is a bit tricky (and different from the original code in this function).  I see what it's doing, but a comment explaining why would be helpful.  Perhaps a reference to where `home_tmp` is created in `take_incr_backup`, and a comment there as well referencing this.

File: test/suite/wtbackup.py [hunk 12]
```diff
@@ -124,17 +128,68 @@ def confirmPathDoesNotExist(self, uri, dir):
...
+
+    #
+    # Cursor next wrapper function.
+    #
+    def next(self, cur):
+        ret = cur.next()
+        if ret == 0:
+           return True
+        self.assertEqual(ret, wiredtiger.WT_NOTFOUND)
+        return False
```
Comment (ddanderson) [inline test/suite/wtbackup.py:151]: I steered you wrong on this, I think we don't need this function and code can just say: `while cursor.next() == 0:`.
The reason is that the Python layer normally takes care of any errors from any WiredTiger API.  All errors are converted to exceptions (which would fail the Python test).  And the API understands that cursor.next() and some other functions can legally return `WT_NOTFOUND` without error.  So `cursor.next` can only return `0` or `WT_NOTFOUND`, or not return at all and raise an an exception instead.

**Review by ddanderson — APPROVED** [review]
LGTM!  This is looking really great, it makes sense and code size is greatly reduced.  I tagged some smaller issues that could be improved, but don't need a final review.

**Review by sueloverso — APPROVED** [review]
LGTM

File: test/suite/wtbackup.py (hunk 12 above)
Comment (boboeee) [reply test/suite/wtbackup.py:151]: ahh makes alot of sense, changed back

File: test/suite/wtbackup.py (hunk 11 above)
Comment (boboeee) [reply test/suite/wtbackup.py:114]: done!

File: test/suite/wtbackup.py (hunk 10 above)
Comment (boboeee) [reply test/suite/wtbackup.py:102]: done!
